sound/soc/sof/sof-audio.c

```diff
@@ -250,6 +250,296 @@ static int sof_machine_check(struct snd_sof_dev *sdev)
+		struct sof_ipc_reply reply;
+		struct sof_ipc_dai_config *config = dai->dai_config;
+
+		if (!config) {
+			dev_err(dev, "error: no config for DAI %s\n",
+				dai->name);
+			continue;
+		}
+
+		/*
+		 * The link DMA channel would be invalidated for running
+		 * streams but not for streams that were in the PAUSED
+		 * state during suspend. So invalidate it here before setting
+		 * the dai config in the DSP.
+		 */
+		if (config->type == SOF_DAI_INTEL_HDA)
+			config->hda.link_dma_ch = DMA_CHAN_INVALID;
+
+		ret = sof_client_tx_message(scomp->dev,
+					    config->hdr.cmd, config,
+					    config->hdr.size,
+					    &reply, sizeof(reply));
+
+		if (ret < 0) {
+			dev_err(dev,
+				"error: failed to set dai config for %s\n",
+				dai->name);
+
+			return ret;
+		}
+	}
+
+	/* complete pipeline */
+	list_for_each_entry(swidget, &sof_audio->widget_list, list) {
+		switch (swidget->id) {
+		case snd_soc_dapm_scheduler:
+			swidget->complete =
+				snd_sof_complete_pipeline(scomp, swidget);
+			break;
+		default:
+			break;
+		}
+	}
+
+	/* restore pipeline kcontrols */
+	ret = sof_restore_kcontrols(dev);
+	if (ret < 0)
+		dev_err(dev,
+			"error: restoring kcontrols after resume\n");
+
+	return ret;
+}
+
+static int sof_destroy_pipelines(struct device *dev)
+{
+	struct sof_audio_dev *sof_audio = sof_get_client_data(dev);
+	struct snd_sof_widget *swidget;
+	struct sof_ipc_free ipc_free;
+	struct sof_ipc_reply reply;
+	int ret = 0;
+
+	list_for_each_entry_reverse(swidget, &sof_audio->widget_list, list) {
+		memset(&ipc_free, 0, sizeof(ipc_free));
+
+		/* skip if there is no private data */
+		if (!swidget->private)
+			continue;
+
+		/* configure ipc free message */
+		ipc_free.hdr.size = sizeof(ipc_free);
+		ipc_free.hdr.cmd = SOF_IPC_GLB_TPLG_MSG;
+		ipc_free.id = swidget->comp_id;
+
+		switch (swidget->id) {
+		case snd_soc_dapm_scheduler:
+			ipc_free.hdr.cmd |= SOF_IPC_TPLG_PIPE_FREE;
+			break;
+		case snd_soc_dapm_buffer:
+			ipc_free.hdr.cmd |= SOF_IPC_TPLG_BUFFER_FREE;
+			break;
+		default:
+			ipc_free.hdr.cmd |= SOF_IPC_TPLG_COMP_FREE;
+			break;
+		}
+		ret = sof_client_tx_message(dev, ipc_free.hdr.cmd,
+					    &ipc_free, sizeof(ipc_free), &reply,
+					    sizeof(reply));
+		if (ret < 0) {
+			dev_err(dev,
+				"error: failed to free widget type %d with ID: %d\n",
+				swidget->widget->id, swidget->comp_id);
+
+			return ret;
+		}
+	}
+
+	return ret;
+}
+
+static int sof_audio_resume(struct device *dev)
+{
+	return sof_restore_pipelines(dev);
+}
+
+static int sof_audio_suspend(struct device *dev)
+{
+	return sof_set_hw_params_upon_resume(dev);
+}
+
+static int sof_audio_runtime_suspend(struct device *dev)
+{
+	return sof_destroy_pipelines(dev);
+}
+
+static const struct dev_pm_ops sof_audio_pm = {
+	SET_SYSTEM_SLEEP_PM_OPS(sof_audio_suspend, sof_audio_resume)
+	SET_RUNTIME_PM_OPS(sof_audio_runtime_suspend, sof_audio_resume, NULL)
+};
+
 static int sof_audio_probe(struct platform_device *pdev)
```

Comment:
@ranj063 I'm having a problem with sof_audio_resume being called without a prior call to sof_audio_runtime_suspend. This means that sof_destroy_pipelines is never called so when sof_restore_pipelines is called there are a lot of problem.

Any, idea why this might happen? It looks like sof_audio_suspend and sof_audio_resume are not symmetrical, so basically if runtime PM is disabled this looks like trouble.

Reply:
Do you mean

```
--- a/sound/soc/sof/sof-of-dev.c
+++ b/sound/soc/sof/sof-of-dev.c
@@ -39,6 +39,7 @@ static void sof_of_probe_complete(struct device *dev)
        /* allow runtime_pm */
        pm_runtime_set_autosuspend_delay(dev, SND_SOF_SUSPEND_DELAY_MS);
        pm_runtime_use_autosuspend(dev);
+       pm_runtime_set_active(dev);
        pm_runtime_enable(dev);
 }
 ```
Let me give it a try. Also, I noticed now that I'm commenting on the older PR. But the things is still a fact for the your current PR.

Reply:
@ranj063 the above change doesn't seem to work, but i have tons of hacks. Will need to investigate more, perhaps you can also point me to the similar change you did for acpi.

Reply:
> Let me give it a try. Also, I noticed now that I'm commenting on the older PR. But the things is still a fact for the your current PR.

Yes, I have fixed it my current PR.

Reply:
I see, thanks will close this comment and put any other discussion on the current PR.

Reply:
@ranj063 Ok this works